File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    answer = []
    dr = [0, 1, 0, -1]
    dc = [1, 0, -1, 0]
    num_map = [[r * columns + c for c in range(1, columns + 1)] for r in range(rows)] 
    
    for query in queries:
        # l_ 최소 row, column, m_최대 row, column
        lr, lc, mr, mc = query[0] - 1, query[1] - 1, query[2] - 1, query[3] - 1
        min_num = num_map[lr][lc]
        num_map[lr][lc], temp = num_map[lr + 1][lc], num_map[lr][lc]
        
        # 시작좌표(temp에 시작좌표를 넣기 때문에 시작은 한칸 옆에서 시작), 방향
        r, c, d = lr, lc + 1, 0
        
        # 최소 좌표로 오면 회전 끝
        while r != lr or c != lc:
            # 최소값 갱신
            if min_num > num_map[r][c]:
                min_num = num_map[r][c]
            num_map[r][c], temp = temp, num_map[r][c]
            nr, nc = r + dr[d], c + dc[d]
            if lr <= nr <= mr and lc <= nc <= mc:
                r, c = nr, nc
            else:
                d += 1
                r, c = r + dr[d], c + dc[d]
        answer.append(min_num)
    return answer

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Arrays;

class Solution {
    static int[][] numbers;
    static int[] dr = {0, 1, 0, -1};
    static int[] dc = {1, 0, -1, 0};
    
    public int rotation(int sr, int sc, int er, int ec) {
        boolean[][] visited = new boolean[er + 1][ec + 1];
        int minNum = numbers[sr][sc];
        int beforeNum = numbers[sr][sc];
        Deque<int[]> dq = new ArrayDeque<int[]>();
        dq.offer(new int[]{sr, sc});
        
        for (int d = 0; d < 4; d++) {
            while (!dq.isEmpty()) {
                int[] pos = dq.poll();
                int nr = pos[0] + dr[d], nc = pos[1] + dc[d];
                if (sr <= nr && nr <= er && sc <= nc && nc <= ec && !visited[nr][nc]) {
                    int temp = numbers[nr][nc];
                    if (minNum > temp) minNum = temp;
                    numbers[nr][nc] = beforeNum;
                    beforeNum = temp;
                    visited[nr][nc] = true;
                    dq.offer(new int[]{nr, nc});
                } else {
                    dq.offer(pos);
                    break;
                }
            }
        }
        
        return minNum;
    }
    
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];
        
        // 1 ~ rows * columns 숫자로 채워진 배열 만들기
        numbers = new int[rows][columns];
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < columns; c++) {
                numbers[r][c] = r * columns + c + 1;
            }
        }
        
        // 회전 수행
        for (int i = 0; i < queries.length; i++) {
            answer[i] = rotation(queries[i][0] - 1, queries[i][1] - 1, queries[i][2] - 1, queries[i][3] - 1);
        }
        
        return answer;
    }
}